### ISSUE TITLE
No "free" transitive packages

### ISSUE THREAD

**Issue by gzak** [NuGet/Home#3052]
Not sure which repo to put this under, but here goes:

Suppose package X has a dependency on package Y, and I add a dependency just on package X in my application. Package Y will automatically be downloaded and added to my application, but I never had to declare it in my project.json file. This means I'll have access to classes in Y "for free".

This seems convenient, but it can become problematic when a later version of X drops its dependence on Y. When I upgrade to the newest version of X I suddenly get lots of red squiggles in the code, with no hint that I should install Y to resolve the issue.

One solution is to stop giving access to transitive packages "for free", and require applications to add an explicit dependency to the transitive package if they need to access its classes. After all, if they're using classes in Y, it is a direct dependence and no longer just transitive.

This is how traditional references work today: if assembly A has a method which returns an object whose type is defined in assembly B, you have to also add a reference to assembly B before you can use it.

**Comment by rrelyea** [comment]
When you upgrade from v1 of X to v2 of X, does the NuGet VS UI tell you of the dependency changes? If so, can you paste in bitmap? If not, good to consider for future.

**Comment by gzak** [comment]
The NuGet VS UI does tell you of the dependency changes:

```
Updating 'X' from version '0.0.1' to '0.0.2' in project 'MyProject'.
Removing 'X 0.0.1' from MyProject.
Successfully removed 'X 0.0.1' from MyProject.
Removing 'Y 0.0.1' from MyProject.
Successfully removed 'Y 0.0.1' from MyProject.
Adding 'X 0.0.2' to MyProject.
Installing 'X 0.0.2'.
Successfully installed 'X 0.0.2'.
Successfully added 'X 0.0.2' to MyProject.
Uninstalling 'X 0.0.1'.
Successfully uninstalled 'X 0.0.1'.
Uninstalling 'Y 0.0.1'.
Successfully uninstalled 'Y 0.0.1'.
```

**Comment by davidfowl** [comment]
You can control this on a per dependency basis when authors your package. We can't break transitive compilation references as it would be a huge breaking change.

**Comment by gzak** [comment]
@davidfowl mind clarifying how to do this? Link to docs would be sufficient. Thanks!

**Comment by yishaigalatzer** [comment]
@davidfowl I think @gzak is referring to the packages.config scenario.

This is the intended design, it allows for a package author to create what we call a meta-package that brings in multiple dependencies with one reference.

The downside @gzak describes is real, but it is also rare. The design favors the common scenario over the rare one, and dropping the transitive nature of NuGet will cause more pain than (in my opinion) value. 

For example: There are too many times where the top level package exposes types on it's APIs that use types from the dependent package. Dropping the transitive nature of NuGet will lead to broken experience by default when installing that kind of package.

Another example: Keeping this structure allows for very useful refactoring of individual packages. We just refactoring nuget.server to a core package and an "IIS" package, that allows the existing consumers to just update their package without any impact, but enabled a scenario for self hosting nuget.server for people that wanted it.

Last a package author that drops a dependency is in essence doing a breaking change, similar to dropping an API from his package and moving it to another one.

So in conclusion although I agree that the promblem statement is correct and accurate, the downsides to taking the suggestion are bigger than the value and hence I would recommend against taking a fix here.

**Comment by davidfowl** [comment]
Ah with `packages.config` there's no real way around this...

**Comment by JanReimerD** [comment]
Hi,

I want to pick this up since we've been thinking about packaging sceanrios a lot. In my eyes what NuGet needs is the feature to allow "Runtime Dependencies". What's been stated here is correct: One needs to be able to state "My package needs package xyz at version a.b.c.". Even if xyz is not used in the API but as technology for the implementation this dependency is required because the corresponding DLLs must be there during runtime. This is vital for the whole packaging system since a version mismatch must be detected and consolidated.

However, the downside of the current approach is that all transitive dependencies are added to the project as reference. Imagine that xyz has futher dependencies which have again further dependencies. This totally breaks the principle of information hiding. This is the key point! NuGet beaks this basic architecture pattern! The client of package xyz does not want to use those transitive dependencies, maybe he even is not meant to use those technologies directly. However, NuGet forces the client to have them all as reference.

A runtime dependency would solve this problem. The author of package xyz could state "I need package xyz in version a.b.c at runtime." Nuget would download those packages, put the binaries into the target folder but won't add a project reference to the client of xyz. Maybe a post build copy step must be added to the project in order to achieve this. It's up to the author of package xyz to decide whether his package needs a regular dependency (since it is part of the API) or just a runtime dependency.

We had this issue while packaging native C++ projects. We have scenarios where we want to hide a complete system with lots of components behind a service interface. The user of the system has dependencies to the service package but it should not link directly to the underlying implementation (which can be tens of packages). We solved this by splitting .h/.lib files and .dll files into two packages. The service package has NuGet-Dependencies to .dll-Packages only. This doesn't work for C# since there is no difference between runtime lib (dll) and symbol lib with header (.h/.lib) like in C++.

With .Net Core this problem is growing in my eyes since everything is split into tiny little packages with lots of transitive dependencies.

Best regards
Jan

**Comment by davidfowl** [comment]
It's already supported in nuget v3 see https://github.com/NuGet/Home/wiki/%5BSpec%5D-Managing-dependency-package-assets
